File: Signal/test/SignalBaseTest.m
//
// Copyright 2018 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "SignalBaseTest.h"
#import <SignalMessaging/SignalMessaging-Swift.h>
#import <SignalServiceKit/SDSDatabaseStorage+Objc.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>
#import <SignalServiceKit/TestAppContext.h>

NS_ASSUME_NONNULL_BEGIN

@implementation SignalBaseTest

- (void)setUp
{
    [super setUp];

    SetCurrentAppContext([TestAppContext new], YES);
    [MockSSKEnvironment activate];
    [MockSMEnvironment activate];

    [SSKEnvironment.shared setGroupsV2ForUnitTests:[GroupsV2Impl new]];
}

- (void)tearDown
{
    [MockSSKEnvironment flushAndWait];
    [super tearDown];
}

- (void)readWithBlock:(void (^NS_NOESCAPE)(SDSAnyReadTransaction *))block
{
    [SDSDatabaseStorage.shared readWithBlock:block];
}

- (void)writeWithBlock:(void (^NS_NOESCAPE)(SDSAnyWriteTransaction *))block
{
    DatabaseStorageWrite(SDSDatabaseStorage.shared, block);
}

@end

NS_ASSUME_NONNULL_END

File: Signal/test/SignalBaseTest.swift
//
// Copyright 2020 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

public import XCTest

@testable import SignalServiceKit

open class SignalBaseTest: XCTestCase {
    private var oldContext: (any AppContext)!

    @MainActor
    override public func setUp() {
        super.setUp()
        let setupExpectation = expectation(description: "mock ssk environment setup completed")
        self.oldContext = CurrentAppContext()
        Task {
            await MockSSKEnvironment.activate()
            setupExpectation.fulfill()
        }
        waitForExpectations(timeout: 30)
    }

    @MainActor
    override open func tearDown() {
        MockSSKEnvironment.deactivate(oldContext: self.oldContext)
        super.tearDown()
    }

    func read<T>(block: (DBReadTransaction) throws -> T) rethrows -> T {
        return try SSKEnvironment.shared.databaseStorageRef.read(block: block)
    }

    func write<T>(block: (DBWriteTransaction) throws -> T) rethrows -> T {
        return try SSKEnvironment.shared.databaseStorageRef.write(block: block)
    }
}
